src/trace/entry.zig

```diff
@@ -1,245 +1,83 @@
 const std = @import("std");
-const logfmt = @import("logfmt.zig");
-const Level = @import("level.zig").Level;
-const Channel = @import("../sync/channel.zig").Channel;
+const sig = @import("../sig.zig");
+const trace = @import("lib.zig");
+
+const Level = trace.level.Level;
+const ScopedLogger = trace.log.ScopedLogger;
 const AtomicBool = std.atomic.Value(bool);
 
-pub const Entry = union(enum) {
-    channel_print: ChannelPrintEntry,
-    direct_print: DirectPrintEntry,
-    noop,
+pub fn NewEntry(comptime scope: ?[]const u8) type {
+    return Entry(struct {}, scope);
+}
 
-    const Self = @This();
+pub fn Entry(comptime Fields: type, comptime scope: ?[]const u8) type {
+    return struct {
+        logger: ScopedLogger(scope),
+        level: Level,
+        fields: Fields,
 
-    pub fn init(allocator: std.mem.Allocator, channel: *Channel(logfmt.LogMsg), log_level: Level) Self {
-        return .{ .channel_print = ChannelPrintEntry.init(allocator, channel, log_level) };
-    }
+        const Self = @This();
 
-    pub fn deinit(self: Self) void {
-        switch (self) {
-            .noop => {},
-            .channel_print => |impl| impl.deinit(),
-            .direct_print => |impl| impl.deinit(),
+        pub inline fn init(logger: ScopedLogger(scope), level: Level) Self {
+            return .{ .logger = logger, .level = level, .fields = .{} };
         }
-    }
 
-    pub fn field(self: Self, name: []const u8, value: anytype) Self {
-        switch (self) {
-            .noop => {
-                return self;
-            },
-            .channel_print => |entry| {
-                var log_entry = entry;
-                _ = log_entry.field(name, value);
-                return Entry{ .channel_print = log_entry };
-            },
-            .direct_print => |entry| {
-                var log_entry = entry;
-                _ = log_entry.field(name, value);
-                return Entry{ .direct_print = log_entry };
-            },
+        /// Add a field to the log message.
+        pub inline fn field(
```

Comment:
Shouldn't be `inline`.

Reply:
There is actually a benefit to this being `inline`: if `value` is a comptime-known or comptime-only value (ie consider `.field(metric_name, INT_CONSTANT).log("Beginning work")`), the `inline` here will allow us to retain that comptime-knowness through the call, allowing the value to be directly embedded as type information in a comptime field, rather than taking up space on the stack, achieving the same properties as we would get if we did the log in a single call.

Reply:
I don't see howargument makes sense. `new_fields` already cannot be a "comptime-only" struct given that it copies values from `self.fields`. The copy of `value` into `new_fields` will not benefit at all from the comptime-ness of `value`.

Secondly, we need to stop thinking in terms of "space on the stack." This is an irrelevant idea that should have no say in reasoning about inline functions. It's not "bad" to take up stack space. It's not "bad" to have extra stack frames. These things take orders of magnitude less time than half the logic surrounding them not to mention the fact that the optimizer will easily see over-allocated stack partitions and remove them.

Reply:
> I don't see how this argument makes sense. new_fields already cannot be a "comptime-only" struct given that it copies values from self.fields. The copy of value into new_fields will not benefit at all from the comptime-ness of value.

I'm not saying it becomes comptime-only, I am saying it will gain comptime fields. This runtime struct can perfectly well contain a `comptime` field.

To the rest: I guess talking about taking up space on the stack was the wrong framing; my principle concern is about the information the compiler has when reasoning about the code. If we did `logger.infof(fields, fmt_str, args)`, the compiler would have available a greater amount of information available to it by virtue of `.{ .field = 120 }` being encoded into the type as `struct { comptime field: comptime_int = 120 }`. With this inlined, we retain that property despite our current choice of API (although I just noticed for this to work, we'd actually need to take this into account in the `FieldsPlus` logic as well @dnut).

Though we could also say we just don't care about this information being conveyed to the compiler, uninline it and call it a day, I'm just as well in favor of that.

Reply:
> I'm not saying it becomes comptime-only, I am saying it will gain comptime fields. This runtime struct can perfectly well contain a comptime field.

As you pointed out, we don't make it one, nor should we. https://github.com/Syndica/sig/pull/307/files#diff-d4c236e79e7de5edd2999e5ecbe1cc72433b23964fa8c6e40df19926cad8bc0eR70.
I am strongly against using any type of tuple is_comptime hack here. It's unnecessary and is extremely ugly.

> the compiler would have available a greater amount of information available to it by virtue of .{ .field = 120 } being encoded into the type as struct { comptime field: comptime_int = 120 }

Given the line I linked above, this is an optimization that LLVM itself would need to make, and I'm certain it will. It also, again, does not matter if it does.

Reply:
> we'd actually need to take this into account in the FieldsPlus logic as well @dnut).

Yeah, due to this, I don't think it's worth the complexity.

FieldsPlus would need a parameter to embed the default value into the type, and the `value: anytype` parameter of `field` would need to be conditionally included in the function's return type depending on whether its value is available comptime. Here's what I came up with:

```diff
diff --git a/src/trace/entry.zig b/src/trace/entry.zig
index c4b00ea..70c042c 100644
--- a/src/trace/entry.zig
+++ b/src/trace/entry.zig
@@ -23,13 +23,21 @@ pub fn Entry(comptime Fields: type, comptime scope: ?[]const u8) type {
             self: Self,
             comptime name: [:0]const u8,
             value: anytype,
-        ) Entry(FieldsPlus(name, @TypeOf(value)), scope) {
+        ) Entry(FieldsPlus(
+            name,
+            @TypeOf(value),
+            if (isComptime(@TypeOf(value))) value else null,
+        ), scope) {
             if (self.logger == .noop) return .{
                 .logger = .noop,
                 .level = undefined,
                 .fields = undefined,
             };
-            var new_fields: FieldsPlus(name, @TypeOf(value)) = undefined;
+            var new_fields: FieldsPlus(
+                name,
+                @TypeOf(value),
+                if (isComptime(@TypeOf(value))) value else null,
+            ) = undefined;
             inline for (@typeInfo(Fields).Struct.fields) |existing_field| {
                 @field(new_fields, existing_field.name) = @field(self.fields, existing_field.name);
             }
@@ -51,19 +59,25 @@ pub fn Entry(comptime Fields: type, comptime scope: ?[]const u8) type {
             self.logger.private_log(self.level, self.fields, fmt, args);
         }
 
+        fn isComptime(comptime T: type) bool {
+            return switch (@typeInfo(T)) {
+                .ComptimeInt, .ComptimeFloat => true,
+                else => false,
+            };
+        }
+
         /// Returns a new struct type based on Fields, just with one more field added.
-        fn FieldsPlus(comptime field_name: [:0]const u8, comptime FieldType: type) type {
+        fn FieldsPlus(
+            comptime field_name: [:0]const u8,
+            comptime FieldType: type,
+            comptime default_value: ?FieldType,
+        ) type {
             const info = @typeInfo(Fields);
-            const ActualFieldType = switch (@typeInfo(FieldType)) {
-                .ComptimeFloat => f64,
-                .ComptimeInt => u64,
-                else => FieldType,
-            };
             const new_fields = info.Struct.fields ++ &[_]std.builtin.Type.StructField{.{
                 .name = field_name,
-                .type = ActualFieldType,
-                .default_value = null,
-                .is_comptime = false,
+                .type = FieldType,
+                .default_value = if (default_value) |dv| @ptrCast(&dv) else null,
+                .is_comptime = isComptime(FieldType),
                 .alignment = @alignOf(FieldType),
             }};
             const new_struct = std.builtin.Type.Struct{
```